File: backend/app/repositories/district_repository.py
from typing import List, Dict, Any
from app.utils.dynamodb import DynamoDBClient

class DistrictRepository:
    def __init__(self, table_name: str = "knowyourmla_districts", client: DynamoDBClient = None):
        if client:
            self.client = client
        else:
            self.client = DynamoDBClient(table_name=table_name)

    def get_all_districts(self) -> List[Dict[str, Any]]:
        """
        Fetches all districts from knowyourmla_districts where SK = METADATA.
        """
        filter_expression = "SK = :sk"
        expression_attribute_values = {":sk": "METADATA"}
        return self.client.scan(
            FilterExpression=filter_expression,
            ExpressionAttributeValues=expression_attribute_values
        )

    def get_district_by_id(self, district_id: str) -> Dict[str, Any]:
        """
        Fetches district metadata from knowyourmla_districts.
        """
        try:
            return self.client.get_item(
                Key={"PK": district_id, "SK": "METADATA"}
            )
        except Exception:
            return None

File: frontend/lib/repositories/district.repository.ts
import { DynamoDBWrapper } from "../dynamodb";

export class DistrictRepository {
  private client: DynamoDBWrapper;

  constructor(tableName: string = process.env.DISTRICTS_TABLE || "knowyourmla_districts") {
    this.client = new DynamoDBWrapper(tableName);
  }

  /**
   * Fetches all districts using MetadataIndex GSI.
   *
   * Uses QueryCommand against MetadataIndex (SK = METADATA) instead of a
   * full-table ScanCommand. The GSI is already provisioned in Terraform.
   * ProjectionExpression limits transfer to only the fields consumed by
   * DistrictService and StateService.
   */
  async getAllDistricts() {
    return this.client.query({
      IndexName: "MetadataIndex",
      KeyConditionExpression: "SK = :sk",
      ExpressionAttributeValues: { ":sk": "METADATA" },
      ProjectionExpression:
        "PK, #name, normalized_name, alias, state_id, total_constituencies, image_url, description",
      ExpressionAttributeNames: {
        "#name": "name",
      },
    });
  }

  /**
   * Fetches a single district by its PK using GetItem.
   *
   * @param districtId - The full PK value, e.g. "DISTRICT#chennai".
   * @returns The district item or null if not found.
   */
  async getDistrictById(districtId: string) {
    try {
      return this.client.get({
        Key: { PK: districtId, SK: "METADATA" },
      });
    } catch (error) {
      console.error(`Error fetching district ${districtId}:`, error);
      return null;
    }
  }
}
